File: 1.1.16/codes/C/rc_plot.c
/*
 * ============================================================
 *  RC Circuit Transient Response — Data Generator
 *  Problem 1.1.16 | EC 2007
 *
 *  Circuit:  10V source -- 20kΩ -- [20kΩ || 4µF]
 *
 *  Results:
 *    Vc(t) = 5 * (1 - e^{-25t})   V
 *    ic(t) = 0.50 * e^{-25t}      mA
 *
 *  What this program does:
 *    1. Computes Vc(t) and ic(t) analytically
 *    2. Writes data to  rc_data.csv
 *    3. Writes a gnuplot script  rc_plot.gp
 *    4. Calls gnuplot to produce  rc_response.png
 *
 *  Compile:   gcc rc_plot.c -o rc_plot -lm
 *  Run:       ./rc_plot
 * ============================================================
 */

/* ── STEP 1 : Include required headers ─────────────────── */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

/* ── STEP 2 : Define circuit constants ─────────────────── */
#define VS      10.0        /* Source voltage           (V)  */
#define R1      20000.0     /* Series resistor          (Ω)  */
#define R2      20000.0     /* Shunt  resistor          (Ω)  */
#define C_F     4e-6        /* Capacitance              (F)  */
#define VC0     0.0         /* Initial capacitor voltage(V)  */

#define N_POINTS 2000       /* Number of data points         */

/* ── STEP 3 : Derived quantities ───────────────────────── */
/* Thévenin resistance : R1 || R2 */
#define RTH     ( (R1 * R2) / (R1 + R2) )

/* Steady-state (final) capacitor voltage */
#define VINF    ( VS * R2 / (R1 + R2) )

/* Time constant τ = Rth * C */
#define TAU     ( RTH * C_F )

/* Initial capacitor current : Vc(0+) = 0 → cap is short */
#define IC0_MA  ( (VINF / RTH) * 1e3 )  /* in mA */

/* ── STEP 4 : Function prototypes ──────────────────────── */
double Vc(double t);
double ic_mA(double t);
void   write_csv(int n, double t_end);
void   write_gnuplot_script(double tau_ms, double vinf,
                            double t_end_ms);
void   print_summary(void);

/* ── STEP 5 : Main function ─────────────────────────────── */
int main(void)
{
    double t_end = 5.0 * TAU;       /* plot for 5τ           */

    printf("\n");
    printf("=============================================\n");
    printf("  RC Circuit Transient Response\n");
    printf("  Problem 1.1.16 | EC 2007\n");
    printf("=============================================\n");
    print_summary();

    /* Write CSV data */
    write_csv(N_POINTS, t_end);

    /* Write gnuplot script */
    write_gnuplot_script(TAU * 1e3, VINF, t_end * 1e3);

    /* Try to call gnuplot (optional — needs gnuplot installed) */
    printf("\n  Attempting to plot with gnuplot...\n");
    int ret = system("gnuplot rc_plot.gp 2>/dev/null");
    if (ret == 0)
        printf("  Plot saved as  rc_response.png\n");
    else
        printf("  gnuplot not found. Run manually:\n");
        printf("    gnuplot rc_plot.gp\n");

    printf("\n  Data saved to   rc_data.csv\n");
    printf("  Script saved to rc_plot.gp\n\n");
    return 0;
}

/* ── STEP 6 : Capacitor voltage equation ─────────────────
 *
 *   Vc(t) = V∞ + [Vc(0+) - V∞] * e^{-t/τ}
 *         = 5  + [0      -  5] * e^{-25t}
 *         = 5 * (1 - e^{-25t})   V
 */
double Vc(double t)
{
    return VINF + (VC0 - VINF) * exp(-t / TAU);
}

/* ── STEP 7 : Capacitor current equation ─────────────────
 *
 *   ic(t) = C * dVc/dt
 *         = C * d/dt [ V∞ (1 - e^{-t/τ}) ]
 *         = C * V∞/τ  * e^{-t/τ}
 *         = (V∞ / Rth) * e^{-25t}          A
 *         = 0.50 * e^{-25t}                mA
 */
double ic_mA(double t)
{
    return (VINF / RTH) * exp(-t / TAU) * 1e3;
}

/* ── STEP 8 : Write CSV data file ─────────────────────── */
void write_csv(int n, double t_end)
{
    FILE *fp = fopen("rc_data.csv", "w");
    if (!fp) { perror("fopen rc_data.csv"); exit(1); }

    /* Header row */
    fprintf(fp, "t_ms,Vc_V,ic_mA\n");

    double dt = t_end / (n - 1);
    for (int i = 0; i < n; i++) {
        double t  = i * dt;
        fprintf(fp, "%.6f,%.8f,%.8f\n",
                t * 1e3,        /* time in ms   */
                Vc(t),          /* Vc in V      */
                ic_mA(t));      /* ic in mA     */
    }
    fclose(fp);
    printf("\n  Written %d rows to rc_data.csv\n", n);
}

/* ── STEP 9 : Write gnuplot script ───────────────────── */
void write_gnuplot_script(double tau_ms, double vinf,
                          double t_end_ms)
{
    FILE *fp = fopen("rc_plot.gp", "w");
    if (!fp) { perror("fopen rc_plot.gp"); exit(1); }

    fprintf(fp,
        "# ── gnuplot script for RC circuit response ──\n"
        "set terminal pngcairo enhanced font 'Sans,11' "
        "size 1400,700 background '#0d1117'\n"
        "set output 'rc_response_c.png'\n"
        "\n"
        "# Dark theme colours\n"
        "set style line 101 lc rgb '#30363d' lt 1 lw 1\n"
        "set border lc rgb '#8b949e'\n"
        "set key textcolor rgb '#e6edf3'\n"
        "set tics textcolor rgb '#8b949e'\n"
        "\n"
        "# Layout: two plots side by side\n"
        "set multiplot layout 1,2 "
        "title 'RC Circuit Transient Response  —  "
        "Problem 1.1.16 (EC 2007)' "
        "font 'Sans Bold,13' textcolor rgb '#e6edf3'\n"
        "\n"
        "# ── LEFT: Vc(t) ──────────────────────────────\n"
        "set title 'Capacitor Voltage  V_C(t)' "
        "textcolor rgb '#e6edf3' font 'Sans Bold,11'\n"
        "set xlabel 'Time (ms)' textcolor rgb '#e6edf3'\n"
        "set ylabel 'V_C  (V)' textcolor rgb '#e6edf3'\n"
        "set grid ls 101\n"
        "set yrange [-0.3:6.0]\n"
        "set xrange [0:%.1f]\n"
        "set object 1 rect from graph 0,0 to graph 1,1 "
        "behind fc rgb '#161b22' fs solid\n"
        "\n"
        "# Tau annotation\n"
        "set arrow 1 from %.2f, 0 to %.2f, %.4f "
        "lc rgb '#3fb950' lw 1.5 dt 2 nohead\n"
        "set label 1 sprintf('τ = %.0f ms\\n%.4f V (63.2%%%%)',%.1f,%.4f) "
        "at %.2f, %.4f tc rgb '#3fb950' font 'Sans,9' left\n"
        "\n"
        "# Vinf dashed line\n"
        "set arrow 2 from 0, %.1f to %.1f, %.1f "
        "lc rgb '#d29922' lw 1.5 dt 2 nohead\n"
        "set label 2 'V_{∞} = %.0f V' "
        "at %.1f, %.3f tc rgb '#d29922' font 'Sans,9' right\n"
        "\n"
        "plot 'rc_data.csv' using 1:2 with lines "
        "lc rgb '#58a6ff' lw 2.5 title 'V_C(t) = 5(1 - e^{-25t})  V'\n"
        "\n"
        "unset arrow 1\nunset arrow 2\n"
        "unset label 1\nunset label 2\n"
        "unset object 1\n"
        "\n"
        "# ── RIGHT: ic(t) ─────────────────────────────\n"
        "set title 'Capacitor Current  i_c(t)' "
        "textcolor rgb '#e6edf3' font 'Sans Bold,11'\n"
        "set xlabel 'Time (ms)' textcolor rgb '#e6edf3'\n"
        "set ylabel 'i_c  (mA)' textcolor rgb '#e6edf3'\n"
        "set yrange [-0.03:0.58]\n"
        "set object 2 rect from graph 0,0 to graph 1,1 "
        "behind fc rgb '#161b22' fs solid\n"
        "\n"
        "set label 3 'i_c(0^+) = 0.50 mA' "
        "at 2, 0.51 tc rgb '#3fb950' font 'Sans,9' left\n"
        "\n"
        "plot 'rc_data.csv' using 1:3 with lines "
        "lc rgb '#f78166' lw 2.5 title 'i_c(t) = 0.50 e^{-25t}  mA'\n"
        "\n"
        "unset multiplot\n",

        /* xrange */   t_end_ms,
        /* arrow1 */   tau_ms, tau_ms, vinf * (1 - exp(-1.0)),
        /* label1 */   tau_ms, vinf * (1 - exp(-1.0)),
                       tau_ms, vinf * (1 - exp(-1.0)),
                       tau_ms + 4, vinf * (1 - exp(-1.0)) + 0.1,
        /* arrow2 */   vinf, t_end_ms, vinf,
        /* label2 */   vinf, t_end_ms - 2, vinf + 0.15
    );

    fclose(fp);
    printf("  Written gnuplot script to rc_plot.gp\n");
}

/* ── STEP 10 : Print parameter summary ─────────────────── */
void print_summary(void)
{
    printf("\n  Circuit Parameters\n");
    printf("  ------------------------------------------\n");
    printf("  Source Voltage     Vs   = %.1f V\n",  VS);
    printf("  Series Resistor    R1   = %.0f kΩ\n", R1 / 1e3);
    printf("  Shunt  Resistor    R2   = %.0f kΩ\n", R2 / 1e3);
    printf("  Capacitance        C    = %.0f µF\n",  C_F * 1e6);
    printf("  ------------------------------------------\n");
    printf("  Thévenin Resistance Rth = %.1f kΩ\n", RTH / 1e3);
    printf("  Steady-State Voltage V∞ = %.1f V\n",  VINF);
    printf("  Time Constant       τ   = %.1f ms\n", TAU * 1e3);
    printf("  Decay Exponent     1/τ  = %.1f s⁻¹\n",1.0 / TAU);
    printf("  ------------------------------------------\n");
    printf("  Vc(t) = %.0f(1 - e^{-25t}) V\n",     VINF);
    printf("  ic(t) = %.2f * e^{-25t}   mA\n",     IC0_MA);
    printf("  ------------------------------------------\n");
    printf("  Answer : Option (a)\n");
    printf("  ------------------------------------------\n");
}

File: 1.1.16/codes/Python/rc_plot.py
"""
============================================================
  RC Circuit Transient Response Plotter
  Problem 1.1.16 | EC 2007
  
  Circuit:  10V source -- 20kΩ -- [20kΩ || 4µF]
  
  Results:
    Vc(t) = 5 (1 - e^{-25t})  V
    ic(t) = 0.50 e^{-25t}     mA
============================================================
"""

# ── STEP 1 : Import libraries ─────────────────────────────
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
from matplotlib.patches import FancyArrowPatch
from matplotlib.ticker import AutoMinorLocator

# ── STEP 2 : Define circuit parameters ───────────────────
Vs   = 10          # Source voltage  (V)
R1   = 20e3        # Series resistor (Ω)
R2   = 20e3        # Shunt  resistor (Ω)
C    = 4e-6        # Capacitor       (F)

Rth  = (R1 * R2) / (R1 + R2)          # Thévenin resistance = 10 kΩ
Vinf = Vs * R2 / (R1 + R2)            # Steady-state voltage = 5 V
tau  = Rth * C                         # Time constant = 40 ms
Vc0  = 0                               # Initial capacitor voltage

print("=" * 48)
print("  RC Circuit — Computed Parameters")
print("=" * 48)
print(f"  Thévenin Resistance  R_th = {Rth/1e3:.1f} kΩ")
print(f"  Steady-State Voltage V∞   = {Vinf:.1f} V")
print(f"  Time Constant        τ    = {tau*1e3:.1f} ms")
print(f"  Decay exponent       1/τ  = {1/tau:.1f} s⁻¹")
print("=" * 48)

# ── STEP 3 : Generate time vector ────────────────────────
#    Plot for 5 time constants (captures ~99% of transient)
t_end = 5 * tau
t = np.linspace(0, t_end, 2000)

# ── STEP 4 : Compute Vc(t) and ic(t) ────────────────────
Vc = Vinf + (Vc0 - Vinf) * np.exp(-t / tau)   # V
ic = C * np.gradient(Vc, t) * 1e3              # mA  (numerical differentiation)
ic_exact = (Vinf / Rth) * np.exp(-t / tau) * 1e3  # mA  (analytical)

# ── STEP 5 : Set up the figure ───────────────────────────
plt.style.use('dark_background')

fig = plt.figure(figsize=(14, 9), facecolor='#0d1117')
fig.suptitle(
    "RC Circuit Transient Response  —  Problem 1.1.16 (EC 2007)",
    fontsize=16, fontweight='bold', color='white',
    y=0.97
)

gs = gridspec.GridSpec(
    2, 2,
    figure=fig,
    hspace=0.45, wspace=0.38,
    left=0.08, right=0.96,
    top=0.90, bottom=0.08
)

AX_COLOR   = '#161b22'
GRID_COLOR = '#30363d'
BLUE       = '#58a6ff'
ORANGE     = '#f78166'
GREEN      = '#3fb950'
GOLD       = '#d29922'
WHITE      = '#e6edf3'
TICK_COLOR = '#8b949e'

def style_ax(ax, title):
    ax.set_facecolor(AX_COLOR)
    for spine in ax.spines.values():
        spine.set_edgecolor(GRID_COLOR)
    ax.tick_params(colors=TICK_COLOR, labelsize=9)
    ax.xaxis.label.set_color(WHITE)
    ax.yaxis.label.set_color(WHITE)
    ax.set_title(title, color=WHITE, fontsize=11, fontweight='bold', pad=8)
    ax.grid(which='major', color=GRID_COLOR, linewidth=0.8, linestyle='--')
    ax.grid(which='minor', color=GRID_COLOR, linewidth=0.3, linestyle=':')
    ax.xaxis.set_minor_locator(AutoMinorLocator(4))
    ax.yaxis.set_minor_locator(AutoMinorLocator(4))

t_ms = t * 1e3   # time in milliseconds for x-axis

# ─────────────────────────────────────────────────────────
#  PLOT 1 : Vc(t)
# ─────────────────────────────────────────────────────────
ax1 = fig.add_subplot(gs[0, 0])
style_ax(ax1, "Capacitor Voltage  $V_C(t)$")

ax1.plot(t_ms, Vc, color=BLUE, linewidth=2.2, label=r'$V_C(t) = 5(1-e^{-25t})\ \mathrm{V}$')

# Asymptote line
ax1.axhline(Vinf, color=GOLD, linewidth=1.2, linestyle='--', alpha=0.7,
            label=f'$V_{{\\infty}} = {Vinf:.0f}$ V  (steady state)')

# Tau marker (63.2%)
Vc_tau = Vinf * (1 - np.exp(-1))
ax1.plot(tau * 1e3, Vc_tau, 'o', color=GREEN, markersize=8, zorder=5)
ax1.annotate(
    f'  $\\tau$ = {tau*1e3:.0f} ms\n  $V_C = {Vc_tau:.2f}$ V  (63.2%)',
    xy=(tau * 1e3, Vc_tau),
    xytext=(tau * 1e3 + 15, Vc_tau - 0.8),
    color=GREEN, fontsize=8.5,
    arrowprops=dict(arrowstyle='->', color=GREEN, lw=1.2)
)

# 5-tau marker
ax1.plot(5 * tau * 1e3, Vc[-1], 's', color=ORANGE, markersize=7, zorder=5)
ax1.annotate(
    f'  $5\\tau$: $V_C \\approx {Vc[-1]:.2f}$ V\n  (99.3% charged)',
    xy=(5 * tau * 1e3, Vc[-1]),
    xytext=(5 * tau * 1e3 - 60, Vc[-1] - 1.2),
    color=ORANGE, fontsize=8.5,
    arrowprops=dict(arrowstyle='->', color=ORANGE, lw=1.2)
)

ax1.set_xlabel("Time  (ms)")
ax1.set_ylabel("$V_C$  (V)")
ax1.set_xlim(0, t_end * 1e3)
ax1.set_ylim(-0.3, 6.0)
ax1.legend(fontsize=8.5, facecolor='#161b22', edgecolor=GRID_COLOR,
           labelcolor=WHITE, loc='lower right')

# ─────────────────────────────────────────────────────────
#  PLOT 2 : ic(t)
# ─────────────────────────────────────────────────────────
ax2 = fig.add_subplot(gs[0, 1])
style_ax(ax2, "Capacitor Current  $i_c(t)$")

ax2.plot(t_ms, ic_exact, color=ORANGE, linewidth=2.2,
         label=r'$i_c(t) = 0.50\,e^{-25t}\ \mathrm{mA}$')

# Initial current marker
ax2.plot(0, ic_exact[0], 'o', color=GREEN, markersize=8, zorder=5)
ax2.annotate(
    f'  $i_c(0^+) = {ic_exact[0]:.2f}$ mA\n  (capacitor = short ckt)',
    xy=(0, ic_exact[0]),
    xytext=(25, ic_exact[0] - 0.08),
    color=GREEN, fontsize=8.5,
    arrowprops=dict(arrowstyle='->', color=GREEN, lw=1.2)
)

# Tau marker
ic_tau = ic_exact[0] * np.exp(-1)
ax2.plot(tau * 1e3, ic_tau, 'o', color=BLUE, markersize=8, zorder=5)
ax2.annotate(
    f'  $i_c(\\tau) = {ic_tau:.4f}$ mA\n  (36.8%)',
    xy=(tau * 1e3, ic_tau),
    xytext=(tau * 1e3 + 12, ic_tau + 0.04),
    color=BLUE, fontsize=8.5,
    arrowprops=dict(arrowstyle='->', color=BLUE, lw=1.2)
)

ax2.axhline(0, color=GOLD, linewidth=1.0, linestyle='--', alpha=0.6,
            label='$i_c(\\infty) = 0$ mA')

ax2.set_xlabel("Time  (ms)")
ax2.set_ylabel("$i_c$  (mA)")
ax2.set_xlim(0, t_end * 1e3)
ax2.set_ylim(-0.03, 0.58)
ax2.legend(fontsize=8.5, facecolor='#161b22', edgecolor=GRID_COLOR,
           labelcolor=WHITE, loc='upper right')

# ─────────────────────────────────────────────────────────
#  PLOT 3 : Both on same axes (overlay)
# ─────────────────────────────────────────────────────────
ax3 = fig.add_subplot(gs[1, 0])
style_ax(ax3, "Overlay: $V_C(t)$ and $i_c(t)$")

color_Vc = BLUE
color_ic = ORANGE

lns1, = ax3.plot(t_ms, Vc, color=color_Vc, linewidth=2.2,
                 label=r'$V_C(t)$  [V]')
ax3.set_ylabel("$V_C$  (V)", color=color_Vc)
ax3.tick_params(axis='y', labelcolor=color_Vc)
ax3.set_xlabel("Time  (ms)")

ax3b = ax3.twinx()
ax3b.set_facecolor(AX_COLOR)
lns2, = ax3b.plot(t_ms, ic_exact, color=color_ic, linewidth=2.2,
                  linestyle='--', label=r'$i_c(t)$  [mA]')
ax3b.set_ylabel("$i_c$  (mA)", color=color_ic)
ax3b.tick_params(axis='y', labelcolor=color_ic)
for spine in ax3b.spines.values():
    spine.set_edgecolor(GRID_COLOR)

lines = [lns1, lns2]
labels = [l.get_label() for l in lines]
ax3.legend(lines, labels, fontsize=8.5, facecolor='#161b22',
           edgecolor=GRID_COLOR, labelcolor=WHITE, loc='center right')
ax3.set_xlim(0, t_end * 1e3)

# ─────────────────────────────────────────────────────────
#  PLOT 4 : Info panel — Circuit parameters summary
# ─────────────────────────────────────────────────────────
ax4 = fig.add_subplot(gs[1, 1])
ax4.set_facecolor(AX_COLOR)
for spine in ax4.spines.values():
    spine.set_edgecolor(GRID_COLOR)
ax4.set_xticks([])
ax4.set_yticks([])
ax4.set_title("Circuit Summary", color=WHITE, fontsize=11,
              fontweight='bold', pad=8)

info_lines = [
    ("Circuit",          "10V  —  20kΩ  —  [20kΩ ∥ 4µF]"),
    ("",                 ""),
    ("Initial Voltage",  "$V_C(0^+) = 0$ V"),
    ("Thévenin Voltage", "$V_{th} = 5$ V"),
    ("Thévenin Resist.", "$R_{th} = 10$ kΩ"),
    ("Time Constant",    "$\\tau = 40$ ms"),
    ("Decay Exponent",   "$1/\\tau = 25$ s⁻¹"),
    ("",                 ""),
    ("Capacitor Voltage","$V_C(t) = 5(1-e^{-25t})$ V"),
    ("Capacitor Current","$i_c(t) = 0.50\\,e^{-25t}$ mA"),
    ("",                 ""),
    ("Answer",           "Option (a)   ✔"),
]

y_pos = 0.95
for label, val in info_lines:
    if label == "":
        y_pos -= 0.04
        continue
    if label == "Answer":
        ax4.text(0.04, y_pos, f"{label}:", color=GREEN,
                 fontsize=9.5, fontweight='bold', transform=ax4.transAxes, va='top')
        ax4.text(0.42, y_pos, val, color=GREEN,
                 fontsize=9.5, fontweight='bold', transform=ax4.transAxes, va='top')
    elif label in ("Capacitor Voltage", "Capacitor Current"):
        ax4.text(0.04, y_pos, f"{label}:", color=GOLD,
                 fontsize=9, transform=ax4.transAxes, va='top')
        ax4.text(0.42, y_pos, val, color=GOLD,
                 fontsize=9, transform=ax4.transAxes, va='top')
    else:
        ax4.text(0.04, y_pos, f"{label}:", color=TICK_COLOR,
                 fontsize=9, transform=ax4.transAxes, va='top')
        ax4.text(0.42, y_pos, val, color=WHITE,
                 fontsize=9, transform=ax4.transAxes, va='top')
    y_pos -= 0.075

# ── STEP 6 : Save and show ────────────────────────────────
plt.savefig("rc_response.png", dpi=180, bbox_inches='tight',
            facecolor=fig.get_facecolor())
print("\n  Plot saved as  rc_response.png")
plt.show()
